MAINT: Handle faulty devices

teddyrendahl opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
## Description
<!--- Describe your changes in detail -->
The `lightpath` wasn't suitably protected against faulty devices. Many of the methods shamelessly called `device.inserted` e.t.c not accepting that these usually involve Channel Access calls that are liable to fail. Now, we only ask for device information in `find_device_state`. This returns a `DeviceState` enum that describes the state. This helps limit the number of Channel Access calls and in turn uniformly handle when these things fail.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Closes #61 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Added a test for a broken device

ZLLentz commented on lightpath/path.py:
```diff
@@ -27,6 +28,43 @@
+    elif _out and not _in:
+        return DeviceState.Removed
+    else:
+        return DeviceState.Unknown
```
This method is :+1: :+1:, great simple solution! Puts all of the error checking in one place and introduces a faulted state for the bad situations and broken code.

Does this mean we should have devices raise exceptions when they know they are in a faulted state?

teddyrendahl replied:
Hmmm. Can you give me an example of device we would want to do this? In my mind, `Faulted` would be reserved for a disconnection issue or something beyond just not being in any of the designated states. (In the `clear` method I opted to not even to try and call `.remove` on faulted devices)

ZLLentz replied:
Ah, so a device that knows it's in a bad state either reports "in" and "out", or neither "in" nor "out"? Maybe I'm tripping over semantics for things like the `Attenuator` which has a status PV that shows faulted in some scenarios.

teddyrendahl replied:
Yeah maybe we should think about this, current logic is:
not in or not out -> "Unknown"
in and out -> "Unknown"
unable to report -> "Faulted"

That second one might want to be "Faulted" . My original thought was that we wanted to distinguish connection issues from just things being in the wrong place. Helpful for operators to be like this is "Unknown" I'm going to try and move this, and "Faulted" this IOC is down we need to call someone

ZLLentz replied:
So our discussion was:
not connected -> disconnected
in and not out -> inserted
out and not in -> removed
not in and not out -> unknown
exception raised -> error
seem reasonable?

ZLLentz commented on lightpath/path.py:
```diff
@@ -253,7 +288,9 @@ def clear(self, wait=False, timeout=None,
-                  if not device.removed and hasattr(device, 'remove')]
+                  if find_device_state(device) in (DeviceState.Inserted,
+                                                   DeviceState.Unknown)
+                  and hasattr(device, 'remove')]
```
I don't like four-line list comprehensions for lack of readability, but that's more a matter of personal taste. This seems to do the job.

ZLLentz commented on lightpath/tests/test_path.py:
```diff
@@ -199,14 +223,14 @@ def test_complex_branching(lcls):
 known_table = """\
 +-------+--------+----------+----------+---------+
-| Name  | Prefix | Position | Beamline | Removed |
+| Name  | Prefix | Position | Beamline |   State |
```
:+1:

ZLLentz commented:
btw the assigned tab still has me there after the review, so mission accomplished

teddyrendahl commented:
Alright. Per discussion here and offilne, I launched the whole gambit of states.  `DeviceState` now is very specific about what we know about the device.
```
    Removed:
        Device is removed from the beamline.

    Inserted:
        Device is inserted into the beamline. This may or may not prevent beam
        from reaching downstream devices.

    Unknown:
        Device is reporting neither an inserted or removed state.

    Inconsistent:
        The device is reporting that is both inserted and removed.

    Disconnected:
        We were unable to determine the state of the device because one or more
        of the relevant signals was not available.

    Error:
        Catch-all state for any errors the device reported when asked for its
        state that were not simply a failure to communicate with signals
```

ZLLentz commented on lightpath/path.py:
```diff
@@ -53,14 +85,27 @@ def find_device_state(device):
+        # Check if this was an error with an EPICS connection
         logger.exception("Unable to determine device state for %r", device)
-        return DeviceState.Faulted
+        return DeviceState.Error
```
Should probably add `TimeoutError` too in case this happens in a signal `get` call, which will try to wait for connection.

The canonical way to check for different error types is:
```
except (DisconnectedError, TimeoutError):
    <stuff>
except Exception:
    <stuff>
```
Rather than calling `isinstance` yourself
```

ZLLentz commented:
:+1:
